### ISSUE TITLE
Unwanted redirect after useLogin()

### ISSUE THREAD

**Issue by larsblumberg** [marmelab/react-admin#6583]
**What you were expecting:**

Calls to login() don't do an implicit redirect when custom authProvider involves a redirect flow.

**What happened instead:**

When calling the login function returned by `useLogin()`, react-admin redirects to the homepage. This causes a noticable flicker in the browser page because shortly after react-admin's redirect our authProvider (using auth0) does it's own redirect.

**Steps to reproduce:**

```
const auth0Client = new Auth0Client(...)
const customAuthProvider = {
  ...
  login: async () => {auth0Client.loginWithRedirect()},
  ...
}

...

const login = useLogin()
...
<button onClick={login}>Login</button>
```

**Suggestion**

Similarly to `authProvider.logout()` where we can return a String for the route to redirect to or `false` to skip redirecting, `authProvider.login()` could have the same return semantics so that `return false` from `authProvider.login()` doesn't redirect.

**Environment**

* React-admin version: 13.7.1
* React version: 17.0.2
* Browser: Firefox

**Comment by larsblumberg** [comment]
Current workaround:

```
const customAuthProvider = {
  login: async () => {
    auth0Client.loginWithRedirect()
    throw new Error("Intentional error to prevent react-admin from redirecting")
  },
}
```

**Comment by fzaninotto** [comment]
Good request idea! We'll gladly accept a PR adding this ability.

**Comment by larsblumberg** [comment]
Thanks for your feedback, I will keep an eye on creating a PR for that. Workaround solves the issue for the time being.

**Comment by fzaninotto** [comment]
Fixed by #7999
